cpp/src/interop/dlpack.cpp

```diff
@@ -151,18 +151,22 @@ std::unique_ptr<table> from_dlpack(DLManagedTensor const* managed_tensor,
 
   // We only support 1D and 2D tensors with some restrictions on layout
   if (tensor.ndim == 1) {
-    // 1D tensors must have dense layout (strides == nullptr <=> dense row-major)
-    CUDF_EXPECTS(nullptr == tensor.strides || tensor.strides[0] == 1,
+    // 1D tensors must have dense layout (strides == nullptr <=> dense layout), or have shape (0,)
+    CUDF_EXPECTS(nullptr == tensor.strides || tensor.strides[0] == 1 || tensor.shape[0] == 0,
                  "from_dlpack of 1D DLTensor only for unit-stride data");
   } else if (tensor.ndim == 2) {
-    // 2D tensors must have column-major layout and the fastest dimension must have dense layout
-    CUDF_EXPECTS((
-                   // 1D tensor reshaped into (N, 1) is fine
-                   tensor.shape[1] == 1 && (nullptr == tensor.strides || tensor.strides[0] == 1))
-                   // General case
-                   || (nullptr != tensor.strides && tensor.strides[0] == 1 &&
-                       tensor.strides[1] >= tensor.shape[0]),
-                 "from_dlpack of 2D DLTensor only for column-major unit-stride data");
+    CUDF_EXPECTS(
+      // Empty array is fine. If ncols == 0 then we get an empty dataframe
+      // irrespective of nrows, which is slightly different behaviour from
+      // cudf.DataFrame(np.empty((3, 0))) because there's no way to communicate
+      // the index information out with a table view if no columns exist.
+      (tensor.shape[0] == 0 || tensor.shape[1] == 0)
```

Comment:
The case where `ncols == 0` in the incoming dataframe means this is not an equivalent way of constructing dataframes:

```
arr = cupy.empty((3, 0), dtype=float)
df1 = cudf.DataFrame(arr)
df2 = cudf.io.dlpack.from_dlpack(arr.__dlpack__())
```
In this instance, `df1.index` is a `RangeIndex(0, 3, 1)`, whereas `df2.index` is a `RangeIndex(0, 0, 1)`.

I don't think this can be fixed without some major reworking though, since a table just wraps up some columns and going from a table to a list of columns in the cython layer just loops over the columns.